### DISCUSSION TITLE
Pointer and address memory of x value

### DISCUSSION THREAD

**Discussion by survuvi** [modular/modular#233] [Q&A] [answered]
Hello, why I have this error when I try to print the address memory of x?

```python

var x : Int = 42  # x must be mutable

let xPtr = Pointer[Int].address_of(x) # get address memory of x

print(xPtr.address_of()) # print address memory of x

print(xPtr.load()) # dereference the pointer and print value of x

```
_error: Expression [9]:21:26: invalid call to 'address_of': callee expects 1 argument, but 0 were specified
    print(xPtr.address_of()) # print address memory of x

/.modular/Kernels/mojo/Stdlib/Pointer.mojo:103:5: function declared here
    fn address_of(inout arg: type) -> Self:_

**Comment by daydun** [comment, marked as answer]
`address_of` is a static method and is not meant to be called on an instance, only on the type itself.

If you want to print the address you probably want to write
```py
var x : Int = 42
let xPtr = Pointer[Int].address_of(x)
print(xPtr.__as_index())
print(xPtr.load())
```
```
94602671945872
42
```

**Reply by survuvi** [reply]
Thank you very much

**Reply by Moosems** [reply]
Out of curiosity, why would someone need the address of a type?

**Reply by survuvi** [reply]
I don't know if that's true, but years ago I read that Linus Torvalds makes a lot of use of pointers and memory addresses into linux source code

**Reply by abhinav-upadhyay** [reply]
> Out of curiosity, why would someone need the address of a type?

A pointer is essentially the address of a variable, it means that the pointer points to the data stored at that address in the memory. There are plenty of uses of pointers. They are heavily used in the implementation of many data structures such as linked lists, binary search trees, graphs etc. For example in case of a node of a linked list, you would want to have pointers to the next node. This allows for easy manipulation of the list. For example if you need to delete a node, you just need to update the next field of the previous node to the address of its next to next node (and free the node which has been deleted). 

Similarly, in C strings and arrays are essentially pointers to a block of memory where the data is contiguously stored. Instead of passing around a complete array to functions, we pass the address of the first element of the array. This saves space on the stack and avoids copying of data.

**Reply by Moosems** [reply]
Which I understand completely. But those often use pointers to variable addresses, why would one need it for the type itself?

**Reply by abhinav-upadhyay** [reply]
You also need to know the type of the data stored at the location pointed to by the pointer so that you can do anything meaningful with it. For example, if it is a 64 byte signed int stored there, or an 8 bit unsigned int. When you dereference the pointer, the bytes need to be interpreted based on the type. 

In case of Mojo they have parameterized it by the type of pointer. For example:
```python
var x : Int = 42
let xPtr = Pointer[Int].address_of(x)
```
So xPtr is a pointer to an Int with type `Pointer[Int]`.

**Reply by survuvi** [reply]
@abhinav-upadhyay pointers could be used for AI embeddings too?

**Reply by abhinav-upadhyay** [reply]
> @abhinav-upadhyay pointers could be used for AI embeddings too?

Yes. An embedding is just a vector of floating point numbers. For efficient operations you would want to use a pointer to represent it.

Let's say the size of the embedding is 100, then you would create a pointer of F32 type and allocate memory for storing 100 values starting at that address. 

```python
fn create_arr(size: Int) -> DTypePointer[DType.f32]:
    let arr = DTypePointer[DType.f32].alloc(size)
    # write random values to the array but this could be your embeddings as well
    rand[DType.f32](arr, size)
    return arr
```

The above example taken and modified from my [article](https://codeconfessions.substack.com/i/120916586/mojo-code-example)

**Reply by survuvi** [reply]
your articles are very interesting, please allow easy copy/paste code instead of images
